File: bindings/haskell/src/Proven/SafeFloat.hs
{- SPDX-License-Identifier: PMPL-1.0-or-later -}
{- Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk> -}

-- | Safe floating-point operations via libproven FFI.
--
-- Division, sqrt, ln, and NaN/infinity detection are performed by the
-- Idris 2 verified core.
module Proven.SafeFloat
  ( safeFloatDiv
  , isFinite
  , isNaN'
  , safeSqrt
  , safeLn
  ) where

import Proven.FFI (c_proven_float_div, c_proven_float_is_finite,
                   c_proven_float_is_nan, c_proven_float_sqrt, c_proven_float_ln)
import Proven.Core (floatResultToMaybe)

-- | Safe floating-point division. Returns Nothing on division by zero.
-- Delegates to @proven_float_div@ in libproven.
safeFloatDiv :: Double -> Double -> IO (Maybe Double)
safeFloatDiv a b = floatResultToMaybe <$>
  c_proven_float_div (realToFrac a) (realToFrac b)

-- | Check if a float is finite (not NaN or infinity).
-- Delegates to @proven_float_is_finite@ in libproven.
isFinite :: Double -> IO Bool
isFinite x = do
  result <- c_proven_float_is_finite (realToFrac x)
  return (result /= 0)

-- | Check if a float is NaN.
-- Delegates to @proven_float_is_nan@ in libproven.
isNaN' :: Double -> IO Bool
isNaN' x = do
  result <- c_proven_float_is_nan (realToFrac x)
  return (result /= 0)

-- | Safe square root. Returns Nothing for negative values.
-- Delegates to @proven_float_sqrt@ in libproven.
safeSqrt :: Double -> IO (Maybe Double)
safeSqrt x = floatResultToMaybe <$>
  c_proven_float_sqrt (realToFrac x)

-- | Safe natural logarithm. Returns Nothing for non-positive values.
-- Delegates to @proven_float_ln@ in libproven.
safeLn :: Double -> IO (Maybe Double)
safeLn x = floatResultToMaybe <$>
  c_proven_float_ln (realToFrac x)

File: bindings/ruby/lib/proven/safe_float.rb
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
# frozen_string_literal: true

# Safe floating-point operations via FFI to libproven.
# ALL computation delegates to Idris 2 compiled code.

module Proven
  module SafeFloat
    class << self
      # Check if a float is finite (not NaN or Infinity).
      #
      # @param value [Float]
      # @return [Boolean, nil]
      def finite?(value)
        FFI.invoke_bool(
          "proven_float_is_finite",
          [Fiddle::TYPE_DOUBLE],
          [value.to_f]
        )
      end

      # Check if value is NaN.
      #
      # @param value [Float]
      # @return [Boolean, nil]
      def nan?(value)
        FFI.invoke_bool(
          "proven_float_is_nan",
          [Fiddle::TYPE_DOUBLE],
          [value.to_f]
        )
      end

      # Safe division. Returns nil for division by zero or non-finite result.
      #
      # @param numerator [Float]
      # @param denominator [Float]
      # @return [Float, nil]
      def div(numerator, denominator)
        result = FFI.invoke_float_result(
          "proven_float_div",
          [Fiddle::TYPE_DOUBLE, Fiddle::TYPE_DOUBLE],
          [numerator.to_f, denominator.to_f]
        )
        return nil unless result
        status, value = result
        status == FFI::STATUS_OK ? value : nil
      end

      # Safe square root (returns nil for negative input).
      #
      # @param value [Float]
      # @return [Float, nil]
      def sqrt(value)
        result = FFI.invoke_float_result(
          "proven_float_sqrt",
          [Fiddle::TYPE_DOUBLE],
          [value.to_f]
        )
        return nil unless result
        status, val = result
        status == FFI::STATUS_OK ? val : nil
      end

      # Safe natural logarithm (returns nil for non-positive input).
      #
      # @param value [Float]
      # @return [Float, nil]
      def ln(value)
        result = FFI.invoke_float_result(
          "proven_float_ln",
          [Fiddle::TYPE_DOUBLE],
          [value.to_f]
        )
        return nil unless result
        status, val = result
        status == FFI::STATUS_OK ? val : nil
      end
    end
  end
end
